compiler/rustc_codegen_ssa/src/glue.rs

```diff
@@ -39,7 +39,10 @@ pub fn size_and_align_of_dst<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
             // The info in this case is the length of the str, so the size is that
             // times the unit size.
             (
-                bx.mul(info.unwrap(), bx.const_usize(unit.size.bytes())),
+                // All type sizes must fit into `usize`, so this multiplication cannot wrap.
+                // FIXME: it may also be the case that all allocation sizes must fit into `isize`,
+                // which means this multiplication also can't signed wrap, and we could use `unchecked_smul`.
+                bx.unchecked_umul(info.unwrap(), bx.const_usize(unit.size.bytes())),
```

Comment:
~It is unclear to me if we enforce that all allocation sizes fit into `isize` or not. The [nomicon](https://doc.rust-lang.org/nomicon/vec/vec-alloc.html) says:~
> ~std limits all allocations to isize::MAX bytes~

~(but what about allocations created outside of std?)~

Ah, `slice::from_raw_parts` [documentation](https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html) seems pretty clear:

> The total size `len * mem::size_of::<T>()` of the slice must be no larger than `isize::MAX`.